amethyst_network/src/server/udp.rs

```diff
@@ -0,0 +1,125 @@
+//! All UDP related logic for getting and sending data out to the other side.
+
+use std::net::SocketAddr;
+use std::sync::mpsc::{self, Receiver, Sender};
+use std::thread;
+
+use super::{
+    ClientEvent, PacketReceiving, PacketSending, ReceiveHandler, SendHandler, ServerSocketEvent,
+};
+use laminar::net::UdpSocket;
+use laminar::NetworkConfig;
+
+/// An UDP receiver, wrapper for starting the UDP-receiving thread.
+pub struct UdpReceiver {
+    // socket used for receiving packets
+    socket: UdpSocket,
+}
+
+impl UdpReceiver {
+    /// Create a new instance of UDP server.
+    pub fn new(addr: SocketAddr) -> UdpReceiver {
+        let socket = UdpSocket::bind(&addr, NetworkConfig::default()).unwrap();
+        UdpReceiver { socket }
+    }
+
+    /// This will run the udp receiver on it's own thread.
+    pub fn run(addr: SocketAddr) -> ReceiveHandler {
+        let mut udp_receiver = UdpReceiver::new(addr);
+
+        // channel used for communicating about received packets.
+        let (tx, rx) = mpsc::channel();
+
+        let thread_handle = thread::spawn(move || {
+            udp_receiver.start_receiving(tx);
+        });
+
+        ReceiveHandler::new(5000, rx, thread_handle)
+    }
+}
+
+impl PacketReceiving for UdpReceiver {
+    // 1. Receive data from socket.
+    // 2. Analyze data.
+    // 3. Check if there are events from laminar.
+    // 3. Notify the receiver with the received data over send channel.
+    fn start_receiving(&mut self, tx: Sender<ServerSocketEvent>) {
+        loop {
+            let result = self.socket.recv();
+            match result {
+                Ok(Some(packet)) => {
+                    if let Err(e) = tx.send(ServerSocketEvent::Packet(packet)) {
+                        warn!("Send channel error. Reason: {:?}", e)
+                    }
+                }
+                Ok(None) => {
+                    if let Err(e) = tx.send(ServerSocketEvent::Empty) {
+                        warn!("Send channel error. Reason: {:?}", e)
+                    }
+                }
+                Err(e) => {
+                    if let Err(e) = tx.send(ServerSocketEvent::Error(e)) {
+                        warn!("Send channel error. Reason: {:?}", e)
+                    }
+                }
+            };
+
+            // iter over client events occurred from within laminar.
+            self.socket.events().into_iter().map(|e| {
```

Comment:
nit but a for_each is better here since it describes that you are not mapping the values but actually just using the values.